### PR TITLE
[Feature:Autograding] Add Haskell image

### PR DESCRIPTION
**Pull request by ION606**
### Please check if the PR fulfills these requirements:

* [x] Tests for the changes have been added/updated (if possible)
* [x] Documentation has been updated/added if relevant

### What is the current behavior?

none

### What is the new behavior?
we can now run haskell files

### Other information?

this is not a breaking change and I tested it using some simple files (I included an example below):

```haskell
-- Function to compute the nth Fibonacci number
fibonacci :: Integer -> Integer
fibonacci n
  | n <= 0    = 0
  | n == 1    = 1
  | otherwise = fibonacci (n - 1) + fibonacci (n - 2)

-- Function to compute the factorial of a given number
factorial :: Integer -> Integer
factorial 0 = 1
factorial n = n * factorial (n - 1)

-- Function to check if a number is prime
isPrime :: Integer -> Bool
isPrime n
  | n <= 1    = False
  | n == 2    = True
  | otherwise = null [ x | x <- [2..isqrt n], n `mod` x == 0]
  where
    isqrt = floor . sqrt . fromIntegral

-- Main function to test the above functions
main :: IO ()
main = do
  putStrLn "Enter a number for Fibonacci calculation:"
  fibInput <- getLine
  let fibNumber = read fibInput :: Integer
  putStrLn ("Fibonacci number at position " ++ show fibNumber ++ " is " ++ show (fibonacci fibNumber))

  putStrLn "\nEnter a number for factorial calculation:"
  factInput <- getLine
  let factNumber = read factInput :: Integer
  putStrLn ("Factorial of " ++ show factNumber ++ " is " ++ show (factorial factNumber))

  putStrLn "\nEnter a number to check if it is prime:"
  primeInput <- getLine
  let primeNumber = read primeInput :: Integer
  putStrLn (show primeNumber ++ (if isPrime primeNumber then " is" else " is not") ++ " a prime number.")
```

### CODE REVIEW

File: dockerfiles/haskell/3.0.0.0/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,17 @@
+FROM ubuntu:22.04
+
+# Install necessary packages for GHC (compiler)
+RUN apt-get update \
+    && apt-get -y --no-install-recommends install \
```
Comment (cjreed121) [inline dockerfiles/haskell/3.0.0.0/Dockerfile:5]: Args in wrong place
```suggestion
    && apt-get install -y --no-install-recommends \
```

File: dockerfiles/haskell/3.0.0.0/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    build-essential \
+    curl \
+    libgmp-dev \
```
Comment (cjreed121) [inline dockerfiles/haskell/3.0.0.0/Dockerfile:8]: Are these 3 necessary for install of ghc and cabal-install? Will apt fetch their deps for them?

File: dockerfiles/haskell/3.0.0.0/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    ghc cabal-install \
+    && rm -rf /var/lib/apt/lists/*
+    
+# Maybe add Stack (another build tool for Haskell), up to Instructor?
+# RUN curl -sSL https://get.haskellstack.org/ | sh
```
Comment (cjreed121) [inline dockerfiles/haskell/3.0.0.0/Dockerfile:13]: I'm not too familiar with Haskell build systems. Would stack and cabal be mutually exclusive for a given assignment?

If so, we could make two tags. One could be `3-cabal` and the other `3-stack`.

**Review by cjreed121 — CHANGES_REQUESTED** [review]

**Comment by cjreed121** [comment]
Edit PR description

s/prolog/haskell/

**Comment by ION606** [comment]
> Edit PR description
> 
> 
> 
> s/prolog/haskell/

oops I think I fixed it now

File: dockerfiles/haskell/3.0.0.0/Dockerfile (hunk 3 above)
Comment (ION606) [reply dockerfiles/haskell/3.0.0.0/Dockerfile:13]: I don't think so, they're used together

File: dockerfiles/haskell/3.0.0.0/Dockerfile (hunk 2 above)
Comment (ION606) [reply dockerfiles/haskell/3.0.0.0/Dockerfile:8]: those were needed for `RUN curl -sSL https://get.haskellstack.org/ | sh` but we don't need them, I'll update them.

File: dockerfiles/haskell/3.0.0.0/Dockerfile (hunk 3 above)
Comment (cjreed121) [reply dockerfiles/haskell/3.0.0.0/Dockerfile:10]: You may not need it but I'm wondering if it's something helpful to include. Based on what I'm seeing online, it appears stack and cabal are their own separate build systems so you wouldn't mix the two.

File: dockerfiles/haskell/3.0.0.0/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,14 @@
+FROM ubuntu:22.04
+
+# Install necessary packages for GHC (compiler)
+RUN apt-get update \
+    && apt-get install -y --no-install-recommends \
+    ghc cabal-install \
```
Comment (cjreed121) [inline dockerfiles/haskell/3.0.0.0/Dockerfile:6]: Can we pin ghc and cabal to specific versions?

**Review by cjreed121 — CHANGES_REQUESTED** [review]

File: dockerfiles/haskell/3.0.0.0/Dockerfile (hunk 4 above)
Comment (ION606) [reply dockerfiles/haskell/3.0.0.0/Dockerfile]: I managed to fix them, but the build process is significantly longer. You can find the images at https://hub.docker.com/repository/docker/ion606/haskell/general if you want  to manually compare

File: dockerfiles/haskell/3.0.0.0/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,25 @@
+FROM ubuntu:22.04
+
+# Install necessary dependencies
+RUN apt-get update \
+    && apt-get install -y --no-install-recommends \
+    apt-utils curl build-essential libffi-dev libffi8 libgmp10 \
+    libgmp-dev libncurses-dev libncurses5 libtinfo5 pkg-config \
+    ca-certificates \
+    && rm -rf /var/lib/apt/lists/*
```
Comment (cjreed121) [inline dockerfiles/haskell/3.0.0.0/Dockerfile:9]: Are any of these not required after install of GHCup, GHC, and Cabal?

**Review by cjreed121 — CHANGES_REQUESTED** [review]
I'm not sure where the tag name of `3.0.0.0` comes from. I'm guessing `8.8.4` might be more relevant since that is the ghc version.

**Review by cjreed121 — CHANGES_REQUESTED** [review]
After some local testing, I noticed the chosen ghc version is not compiled for aarch64. The tests should have caught this but because of an error, it was skipped. Could you pick a different ghc version where it is also compiled for aarch64.

File: dockerfiles/haskell/3.0.0.0/Dockerfile (hunk 5 above)
Comment (ION606) [reply dockerfiles/haskell/3.0.0.0/Dockerfile:9]: 3.0.0.0 is from something deprecated, I'll change it

Comment (cjreed121) [reply dockerfiles/haskell/3.0.0.0/Dockerfile]: Can curl, pkg-config, and ca-certificates be removed after install?

**Comment by ION606** [comment]
> After some local testing, I noticed the chosen ghc version is not compiled for aarch64. The tests should have caught this but because of an error, it was skipped. Could you pick a different ghc version where it is also compiled for aarch64.

Can you please elaborate on this please? Running `docker run --platform=linux/amd64` works fine for me (I put a screenshot below)
image

File: dockerfiles/haskell/3.0.0.0/Dockerfile (hunk 5 above)
Comment (ION606) [reply dockerfiles/haskell/3.0.0.0/Dockerfile]: oh after ghc you mean?

Comment (cjreed121) [reply dockerfiles/haskell/3.0.0.0/Dockerfile]: Yeah

Comment (ION606) [reply dockerfiles/haskell/3.0.0.0/Dockerfile]: I'll try it now :3

**Comment by cjreed121** [comment]
Couple things:
* I'm referring to building a new image not running an existing image.
* I'm referring to aarch64 not amd64.

**Comment by ION606** [comment]
> Couple things:
> 
>     * I'm referring to building a new image not running an existing image.
> 
>     * I'm referring to aarch64 not amd64.

oops my bad, what's the error you're getting? Is it
```
Step 2/8 : RUN apt-get update     && apt-get install -y --no-install-recommends     curl build-essential libffi-dev libffi8 libgmp10     libgmp-dev libtinfo5 pkg-config ca-certificates     && rm -rf /var/lib/apt/lists/*
 ---> [Warning] The requested image's platform (linux/arm64/v8) does not match the detected host platform (linux/amd64/v3) and no specific platform was requested
 ---> Running in ebd9b48099c4
exec /bin/sh: exec format error
The command '/bin/sh -c apt-get update     && apt-get install -y --no-install-recommends     curl build-essential libffi-dev libffi8 libgmp10     libgmp-dev libtinfo5 pkg-config ca-certificates     && rm -rf /var/lib/apt/lists/*' returned a non-zero code: 1
```

**Comment by cjreed121** [comment]
@ION606 please check Zulip and I'll follow up there

**Comment by ION606** [comment]
can you check if this works on aarch64 please?

**Pull request merged**